File: backend/src/integrations/ticketing/ticketing_service.js
/**
 * Ticketing Service & Connector Registry — Phase 14.1
 *
 * Coordinates enterprise ticketing connector dispatch without coupling core risk code.
 * Ensures all tickets enforce the 8 mandatory fields:
 * - asset ID
 * - finding ID
 * - severity
 * - owner
 * - evidence link
 * - remediation
 * - CBOM reference
 * - risk score
 */

const { TicketRequest, TicketResponse } = require("../../domain/contracts");
const { JiraConnector } = require("./jira_connector");
const { ServiceNowConnector } = require("./servicenow_connector");
const { GitHubIssuesConnector } = require("./github_issues_connector");
const { GitLabIssuesConnector } = require("./gitlab_issues_connector");
const { WebhookConnector } = require("./webhook_connector");

class TicketingService {
  constructor() {
    this.connectors = new Map();
  }

  /**
   * Registers a connector instance.
   * @param {BaseTicketingConnector} connector
   */
  registerConnector(connector) {
    if (!connector || !connector.name) {
      throw new Error("Invalid connector: missing name");
    }
    this.connectors.set(connector.name, connector);
    return this;
  }

  /**
   * Unregisters a connector.
   * @param {string} name
   */
  unregisterConnector(name) {
    return this.connectors.delete(name);
  }

  /**
   * Retrieves a registered connector.
   * @param {string} name
   */
  getConnector(name) {
    return this.connectors.get(name);
  }

  /**
   * Lists all registered connectors and their metadata.
   */
  listConnectors() {
    const list = [];
    for (const [name, conn] of this.connectors.entries()) {
      list.push({
        name,
        type: conn.connectorType,
      });
    }
    return list;
  }

  /**
   * Formats a ticket payload without dispatching (dry-run / preview).
   * @param {TicketRequest|Object} rawRequest
   * @param {string} connectorName
   */
  formatTicket(rawRequest, connectorName) {
    const conn = this.getConnector(connectorName);
    if (!conn) {
      throw new Error(`Connector '${connectorName}' is not registered`);
    }
    const ticketRequest = rawRequest instanceof TicketRequest ? rawRequest : new TicketRequest(rawRequest);
    return {
      connectorName,
      connectorType: conn.connectorType,
      payload: conn.formatPayload(ticketRequest),
    };
  }

  /**
   * Creates a ticket on a specific connector.
   * @param {TicketRequest|Object} rawRequest
   * @param {string} connectorName
   * @returns {Promise<TicketResponse>}
   */
  async createTicket(rawRequest, connectorName) {
    const conn = this.getConnector(connectorName);
    if (!conn) {
      throw new Error(`Connector '${connectorName}' is not registered`);
    }
    const ticketRequest = rawRequest instanceof TicketRequest ? rawRequest : new TicketRequest(rawRequest);
    return await conn.createTicket(ticketRequest);
  }

  /**
   * Dispatches ticket creation across multiple connectors.
   * @param {TicketRequest|Object} rawRequest
   * @param {string[]} [connectorNames]
   * @returns {Promise<Array<{ connector: string, success: boolean, ticket?: TicketResponse, error?: string }>>}
   */
  async dispatchMulti(rawRequest, connectorNames = []) {
    const ticketRequest = rawRequest instanceof TicketRequest ? rawRequest : new TicketRequest(rawRequest);
    const targetConnectors = connectorNames.length > 0
      ? connectorNames.map((n) => {
          const c = this.getConnector(n);
          if (!c) throw new Error(`Connector '${n}' not found`);
          return c;
        })
      : Array.from(this.connectors.values());

    if (targetConnectors.length === 0) {
      throw new Error("No ticketing connectors available for dispatch");
    }

    const results = [];
    for (const conn of targetConnectors) {
      try {
        const res = await conn.createTicket(ticketRequest);
        results.push({
          connector: conn.name,
          connectorType: conn.connectorType,
          success: true,
          ticket: res,
        });
      } catch (err) {
        results.push({
          connector: conn.name,
          connectorType: conn.connectorType,
          success: false,
          error: err.message,
        });
      }
    }

    return results;
  }

  /**
   * Maps findings into TicketRequests and creates tickets in batch.
   * @param {Array<Object>} findings
   * @param {Object} options
   * @param {string} [options.defaultOwner]
   * @param {string} [options.cbomRef]
   * @param {string[]} [options.connectorNames]
   */
  async batchCreateFromFindings(findings = [], options = {}) {
    const defaultOwner = options.defaultOwner || "crypto-secops@enterprise.com";
    const defaultCbomRef = options.cbomRef || "urn:ecdat:cbom:latest";
    const connectorNames = options.connectorNames || [];

    const results = [];

    for (const finding of findings) {
      const assetId = finding.asset_id || finding.assetId || finding.file_path || "unknown-asset";
      const findingId = finding.finding_id || finding.findingId || finding.rule_id || finding.id || `find-${Date.now()}`;
      const severity = (finding.severity || "MEDIUM").toUpperCase();
      const owner = finding.owner || defaultOwner;
      const evidenceLink = finding.evidence_link || finding.evidenceLink || (finding.file_path ? `${finding.file_path}#L${finding.line_number || 1}` : "https://ecdat.internal/evidence");
      const remediation = finding.remediation || (finding.developer_feedback && finding.developer_feedback.safe_fix && finding.developer_feedback.safe_fix.summary) || "Upgrade algorithm according to enterprise crypto policy.";
      const cbomRef = finding.cbom_ref || finding.cbomRef || defaultCbomRef;
      const riskScore = finding.risk_score !== undefined
        ? finding.risk_score
        : (finding.riskScore !== undefined ? finding.riskScore : (severity === "CRITICAL" ? 9.5 : 7.0));

      const ticketRequest = new TicketRequest({
        assetId,
        findingId,
        severity,
        owner,
        evidenceLink,
        remediation,
        cbomRef,
        riskScore,
        title: finding.title,
        description: finding.description,
        metadata: finding.metadata || {},
      });

      const dispatchResult = await this.dispatchMulti(ticketRequest, connectorNames);
      results.push({
        findingId,
        assetId,
        dispatch: dispatchResult,
      });
    }

    return {
      totalFindings: findings.length,
      processed: results.length,
      results,
    };
  }
}

// Global default singleton instance
const defaultTicketingService = new TicketingService();

module.exports = {
  TicketingService,
  defaultTicketingService,
  JiraConnector,
  ServiceNowConnector,
  GitHubIssuesConnector,
  GitLabIssuesConnector,
  WebhookConnector,
};

File: scanners/integrations/ticketing_service.py
"""
Ticketing Service (Python) — Phase 14.1

Central registry and dispatcher for enterprise ticketing connectors.
Enforces that core risk code is decoupled from vendor implementations.
"""

from typing import Any, Dict, List, Optional

from scanners.integrations.base_connector import BaseTicketingConnector
from scanners.integrations.ticket_request import TicketRequest, TicketResponse


class TicketingService:
    def __init__(self):
        self._connectors: Dict[str, BaseTicketingConnector] = {}

    def register_connector(self, connector: BaseTicketingConnector) -> "TicketingService":
        if not connector or not connector.name:
            raise ValueError("Connector requires a valid name")
        self._connectors[connector.name] = connector
        return self

    def unregister_connector(self, name: str) -> bool:
        return self._connectors.pop(name, None) is not None

    def get_connector(self, name: str) -> Optional[BaseTicketingConnector]:
        return self._connectors.get(name)

    def list_connectors(self) -> List[Dict[str, str]]:
        return [{"name": name, "type": conn.connector_type} for name, conn in self._connectors.items()]

    def format_ticket(self, raw_request: Any, connector_name: str) -> Dict[str, Any]:
        conn = self.get_connector(connector_name)
        if not conn:
            raise KeyError(f"Connector '{connector_name}' is not registered")

        req = raw_request if isinstance(raw_request, TicketRequest) else TicketRequest(**raw_request)
        return {
            "connector_name": connector_name,
            "connector_type": conn.connector_type,
            "payload": conn.format_payload(req),
        }

    def create_ticket(self, raw_request: Any, connector_name: str) -> TicketResponse:
        conn = self.get_connector(connector_name)
        if not conn:
            raise KeyError(f"Connector '{connector_name}' is not registered")

        return conn.create_ticket(raw_request)

    def dispatch_multi(self, raw_request: Any, connector_names: Optional[List[str]] = None) -> List[Dict[str, Any]]:
        req = raw_request if isinstance(raw_request, TicketRequest) else TicketRequest(**raw_request)

        target_connectors = (
            [self.get_connector(name) for name in connector_names]
            if connector_names
            else list(self._connectors.values())
        )

        if not target_connectors or any(c is None for c in target_connectors):
            raise KeyError("One or more specified connectors were not found or no connectors registered")

        results = []
        for conn in target_connectors:
            assert conn is not None
            try:
                ticket_res = conn.create_ticket(req)
                results.append(
                    {
                        "connector": conn.name,
                        "connector_type": conn.connector_type,
                        "success": True,
                        "ticket": ticket_res.to_dict(),
                    }
                )
            except Exception as e:
                results.append(
                    {
                        "connector": conn.name,
                        "connector_type": conn.connector_type,
                        "success": False,
                        "error": str(e),
                    }
                )
        return results

    def batch_create_from_findings(
        self,
        findings: List[Dict[str, Any]],
        default_owner: str = "crypto-secops@enterprise.com",
        cbom_ref: str = "urn:ecdat:cbom:latest",
        connector_names: Optional[List[str]] = None,
    ) -> Dict[str, Any]:
        results = []
        for finding in findings:
            asset_id = finding.get("asset_id") or finding.get("assetId") or finding.get("file_path") or "unknown-asset"
            finding_id = (
                finding.get("finding_id")
                or finding.get("findingId")
                or finding.get("rule_id")
                or finding.get("id")
                or "find-default"
            )
            severity = str(finding.get("severity", "MEDIUM")).upper()
            owner = finding.get("owner") or default_owner
            evidence_link = (
                finding.get("evidence_link")
                or finding.get("evidenceLink")
                or (
                    f"{finding['file_path']}#L{finding.get('line_number', 1)}"
                    if finding.get("file_path")
                    else "https://ecdat.internal/evidence"
                )
            )
            remediation = (
                finding.get("remediation")
                or (finding.get("developer_feedback", {}).get("safe_fix", {}).get("summary"))
                or "Upgrade algorithm according to enterprise crypto policy."
            )
            c_ref = finding.get("cbom_ref") or finding.get("cbomRef") or cbom_ref
            risk_score = finding.get("risk_score")
            if risk_score is None:
                risk_score = 9.5 if severity == "CRITICAL" else 7.0

            req = TicketRequest(
                asset_id=asset_id,
                finding_id=finding_id,
                severity=severity,
                owner=owner,
                evidence_link=evidence_link,
                remediation=remediation,
                cbom_ref=c_ref,
                risk_score=float(risk_score),
                title=finding.get("title"),
                description=finding.get("description"),
                metadata=finding.get("metadata", {}),
            )

            dispatch_res = self.dispatch_multi(req, connector_names)
            results.append(
                {
                    "finding_id": finding_id,
                    "asset_id": asset_id,
                    "dispatch": dispatch_res,
                }
            )

        return {
            "total_findings": len(findings),
            "processed": len(results),
            "results": results,
        }
